go-controller/pkg/clustermanager/clustermanager.go

```diff
@@ -0,0 +1,271 @@
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"reflect"
+	"sync"
+	"time"
+
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kube"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/metrics"
+	nad "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/network-attach-def-controller"
+	ovntypes "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+// ovnkubeClusterManagerLeaderMetrics object is used for the cluster manager
+// leader metrics
+type ovnkubeClusterManagerLeaderMetrics struct{}
+
+func (ovnkubeClusterManagerLeaderMetrics) On(string) {
+	metrics.MetricClusterManagerLeader.Set(1)
+}
+
+func (ovnkubeClusterManagerLeaderMetrics) Off(string) {
+	metrics.MetricClusterManagerLeader.Set(0)
+}
+
+type ovnkubeClusterManagerLeaderMetricsProvider struct{}
+
+func (_ ovnkubeClusterManagerLeaderMetricsProvider) NewLeaderMetric() leaderelection.SwitchMetric {
+	return ovnkubeClusterManagerLeaderMetrics{}
+}
+
+// ClusterManager structure is the object which manages the cluster nodes.
+// It creates a default network controller for the default network and a
+// secondary network cluster controller manager to manage the multi networks.
+type ClusterManager struct {
+	client                               clientset.Interface
+	defaultNetClusterController          *defaultNetworkClusterController
+	wf                                   *factory.WatchFactory
+	wg                                   *sync.WaitGroup
+	stopChan                             chan struct{}
+	secondaryNetClusterControllerManager *secondaryNetworkClusterControllerManager
+	// event recorder used to post events to k8s
+	recorder record.EventRecorder
+}
+
+// NewClusterManager creates a new Cluster Manager for managing the
+// cluster nodes.
+func NewClusterManager(ovnClient *util.OVNClusterManagerClientset, wf *factory.WatchFactory,
+	wg *sync.WaitGroup, recorder record.EventRecorder) *ClusterManager {
+	cm := &ClusterManager{
+		client:                      ovnClient.KubeClient,
+		defaultNetClusterController: newDefaultNetworkClusterController(ovnClient, wf),
+		wg:                          wg,
+		wf:                          wf,
+		stopChan:                    make(chan struct{}),
+		recorder:                    recorder,
+	}
+
+	if config.OVNKubernetesFeature.EnableMultiNetwork {
+		cm.secondaryNetClusterControllerManager = newSecondaryNetworkClusterControllerManager(ovnClient, wf, recorder)
+	}
+	return cm
+}
+
+// Start waits until this process is the leader before starting master functions
+func (cm *ClusterManager) Start(nodeName string, ctx context.Context) error {
```

Comment:
I think you shouldnt use context and instead should use the stop channel. Pass this stop channel then to `secondaryNetClusterControllerManager.Run(channel)`. The stop chan is being use to stop components where as context is used to close the stop channel. Its the pattern I see in `runOvnkube()`:

```
	stopChan := make(chan struct{})
	wg := &sync.WaitGroup{}
	defer func() {
		close(stopChan)
		wg.Wait()
	}()
```